### ISSUE TITLE
Support 3D pose estimation

### ISSUE THREAD

**Issue by tucan9389** [tucan9389/PoseEstimation-CoreML#35]
## lightweight-human-pose-estimation-3d-demo.pytorch

repo: https://github.com/Daniil-Osokin/lightweight-human-pose-estimation-3d-demo.pytorch

### Models

- pytorch model: [human-pose-estimation-3d.pth.zip](https://github.com/tucan9389/PoseEstimation-CoreML/files/5009665/human-pose-estimation-3d.pth.zip)
- onnx model: [human-pose-estimation-3d.onnx.zip](https://github.com/tucan9389/PoseEstimation-CoreML/files/5009667/human-pose-estimation-3d.onnx.zip)
- mlmodel: [human-pose-estimation-3d.mlmodel.zip](https://github.com/tucan9389/PoseEstimation-CoreML/files/5009664/human-pose-estimation-3d.mlmodel.zip)

### Metadata

<img src="https://user-images.githubusercontent.com/37643248/89092224-ccccc000-d3ea-11ea-8e2b-992614ba8646.png" width=400px>

![image](https://user-images.githubusercontent.com/37643248/89092260-161d0f80-d3eb-11ea-85fb-6312c63f28de.png)

- input: a 256x448 color image
- outputs:
  -

**Comment by sebo361** [comment]
Hi @tucan9389 how about using ARKit's 3D motion tracking: [https://developer.apple.com/documentation/arkit/capturing_body_motion_in_3d](https://developer.apple.com/documentation/arkit/capturing_body_motion_in_3d) ? Or do you think your proposed 3D pose estimation is more accurate?

**Comment by tucan9389** [comment]
@sebo361 
Thanks for your suggestion. I agree with that If the case prefers more. 
But I think there is some other side of benefits when using Core ML rather than AR Kit.

1. When you want to inference person's 3D keypoints with a single image
2. When you want to inference not only a person's but also an object's 3D keypoints which was trained by yourself

**Comment by sebo361** [comment]
@tucan9389 
True, these are interesting benefits to find out! However I am more interested on comparing both approaches regarding the inference speed and accuracy of 3D human pose keypoints. Now as the latest iPad Pro (and iPhone 12) have the LIDAR sensor, i expect to have more precision in 3D human motion tracking, but I am note sure if ARKit's 3D motion tracking uses LIDAR Depth API by default - do you have any insights of that?

Maybe I should setup a project to compare ARKit's 3D motion tracking with your proposed CoreML 3D pose estimation.

**Comment by jookovjook** [comment]
@sebo361
Also, Apple's 3D body tracking `ARBodyTrackingConfiguration` can't work simultaneously with the world `ARWorldTrackingConfiguration`. So, unfortunately, you can't use Apple's 3D body tracking if you also need world tracking

**Comment by tucan9389** [comment]
https://github.com/tucan9389/PoseEstimation-TFLiteSwift

Here is 3D pose estimation demo by using TFLiteSwift. I implemented softargmax with pure Swift and Accelerate framework. In Numpy or Pytorch(or TF), there support dimension summation to use easily, but in swift there is no softargmax function in multi-dimension matrix(tensor). So the implementation of softargmax is a little bit hard. 


<img src="https://user-images.githubusercontent.com/37643248/110994933-e68ca780-83bc-11eb-8331-d827e19d2d36.gif" width=280px>
